File: implementations/go/examples/internal/application/query/get_payments_handler.go
package query

import (
	"context"
	"fmt"

	"github.com/example/account-service/internal/domain/payment"
)

// GetPaymentsQuery queries the list of payment history. No endpoint in this
// repository trusts an owner ID passed by the client — RequesterID is a
// value filled in by the auth middleware, and the list always returns only
// the requester's own payment history (there is no query parameter like
// ?ownerId=).
type GetPaymentsQuery struct {
	RequesterID string
	Page        int
	Take        int
}

type GetPaymentsHandler struct {
	payments payment.Query
}

func NewGetPaymentsHandler(payments payment.Query) *GetPaymentsHandler {
	return &GetPaymentsHandler{payments: payments}
}

func (h *GetPaymentsHandler) Handle(ctx context.Context, q GetPaymentsQuery) (*GetPaymentsResult, error) {
	payments, count, err := h.payments.FindPayments(ctx, payment.FindQuery{
		OwnerID: q.RequesterID,
		Page:    q.Page,
		Take:    q.Take,
	})
	if err != nil {
		return nil, fmt.Errorf("get payments: %w", err)
	}

	results := make([]GetPaymentResult, len(payments))
	for i, p := range payments {
		results[i] = GetPaymentResult{
			PaymentID: p.PaymentID,
			CardID:    p.CardID,
			AccountID: p.AccountID,
			OwnerID:   p.OwnerID,
			Amount:    p.Amount,
			Status:    string(p.Status),
			CreatedAt: p.CreatedAt,
		}
	}
	return &GetPaymentsResult{Payments: results, Count: count}, nil
}

File: implementations/fastapi/examples/src/payment/application/query/get_payments_handler.py
from dataclasses import dataclass

from ...domain.payment_repository import PaymentQuery
from .result import GetPaymentResult, GetPaymentsResult


@dataclass
class GetPaymentsQuery:
    # Looks up only the authenticated requester's own payment history — a client-supplied
    # ownerId is never trusted (no endpoint in this repository accepts a client-specified
    # owner id).
    requester_id: str
    page: int = 0
    take: int = 20


class GetPaymentsHandler:
    def __init__(self, query: PaymentQuery) -> None:
        self._query = query

    async def execute(self, query: GetPaymentsQuery) -> GetPaymentsResult:
        payments, count = await self._query.find_payments(page=query.page, take=query.take, owner_id=query.requester_id)
        return GetPaymentsResult(
            payments=[
                GetPaymentResult(
                    payment_id=p.payment_id,
                    card_id=p.card_id,
                    account_id=p.account_id,
                    owner_id=p.owner_id,
                    amount=p.amount,
                    status=p.status.value,
                    created_at=p.created_at,
                )
                for p in payments
            ],
            count=count,
        )
